File: apps/web/src/data/money.ts
import type { TransactionFlow } from './types';

// Money out is negative, money in is positive. Direction lives in the sign of
// the amount and nowhere else, so an amount and its direction cannot disagree.
export function flowOf(amountMinor: number): TransactionFlow {
  return amountMinor < 0 ? 'DEBIT' : 'CREDIT';
}

export function signedAmount(magnitudeMinor: number, flow: TransactionFlow): number {
  return flow === 'DEBIT' ? -magnitudeMinor : magnitudeMinor;
}

export function magnitudeOf(amountMinor: number): number {
  return Math.abs(amountMinor);
}

// JSON has no negative zero: `-0` serialises as `0`, and the Rust integers on
// the other side of the contract cannot represent it at all. A total that flips
// sign is negated through zero so it never becomes one.
export function negated(amountMinor: number): number {
  return 0 - amountMinor;
}

File: crates/domain/src/money.rs
use crate::types::TransactionFlow;

/// Money out is negative, money in is positive. Direction lives in the sign of
/// the amount and nowhere else, so an amount and its direction cannot disagree.
pub fn flow_of(amount_minor: i64) -> TransactionFlow {
    if amount_minor < 0 {
        TransactionFlow::Debit
    } else {
        TransactionFlow::Credit
    }
}

pub fn signed_amount(magnitude_minor: i64, flow: TransactionFlow) -> i64 {
    match flow {
        TransactionFlow::Debit => -magnitude_minor,
        TransactionFlow::Credit => magnitude_minor,
    }
}

pub fn magnitude_of(amount_minor: i64) -> i64 {
    amount_minor.abs()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn direction_lives_in_the_sign() {
        assert_eq!(flow_of(-1), TransactionFlow::Debit);
        assert_eq!(flow_of(0), TransactionFlow::Credit);
        assert_eq!(flow_of(1), TransactionFlow::Credit);
        assert_eq!(signed_amount(4_325, TransactionFlow::Debit), -4_325);
        assert_eq!(magnitude_of(-4_325), 4_325);
    }
}
